### PR TITLE
[sdk] deduplicate python error logs

### LINKED ISSUE #10222: [sdk/python] simplify python program errors
## Hello!

- Vote on this issue by adding a 👍 reaction
- If you want to implement this feature, comment to let us know (we'll work with you on design, scheduling, etc.)

## Issue details

We currently report 3 errors from Python programs that throw an exception. We should report only 1.

### Reproduce

To reproduce `pulumi new python` and:

```python
import pulumi

raise Exception('!')
```

```pulumi preview```

### Current Behavior

```
Previewing update (dev)

View Live: https://app.pulumi.com/t0yv0/python-err/dev/previews/e08cb126-25ea-4c38-9640-15e381ebc32d

     Type                 Name            Plan       Info
 +   pulumi:pulumi:Stack  python-err-dev  create     3 errors
 
Diagnostics:
  pulumi:pulumi:Stack (python-err-dev):
    error: Program failed with an unhandled exception:
    error: Traceback (most recent call last):
      File "/Users/anton/.nix-profile/bin/pulumi-language-python-exec", line 107, in <module>
        loop.run_until_complete(coro)
      File "/nix/store/7q3qsarzyffcm2sjb2yhjgy0d3hzdi7y-python3-3.8.8/lib/python3.8/asyncio/base_events.py", line 616, in run_until_complete
        return future.result()
      File "/Users/anton/tmp/48h/python-err/venv/lib/python3.8/site-packages/pulumi/runtime/stack.py", line 126, in run_in_stack
        await run_pulumi_func(lambda: Stack(func))
      File "/Users/anton/tmp/48h/python-err/venv/lib/python3.8/site-packages/pulumi/runtime/stack.py", line 49, in run_pulumi_func
        func()
      File "/Users/anton/tmp/48h/python-err/venv/lib/python3.8/site-packages/pulumi/runtime/stack.py", line 126, in <lambda>
        await run_pulumi_func(lambda: Stack(func))
      File "/Users/anton/tmp/48h/python-err/venv/lib/python3.8/site-packages/pulumi/runtime/stack.py", line 149, in __init__
        func()
      File "/Users/anton/.nix-profile/bin/pulumi-language-python-exec", line 106, in <lambda>
        coro = pulumi.runtime.run_in_stack(lambda: runpy.run_path(args.PROGRAM, run_name='__main__'))
      File "/nix/store/7q3qsarzyffcm2sjb2yhjgy0d3hzdi7y-python3-3.8.8/lib/python3.8/runpy.py", line 282, in run_path
        return _run_code(code, mod_globals, init_globals,
      File "/nix/store/7q3qsarzyffcm2sjb2yhjgy0d3hzdi7y-python3-3.8.8/lib/python3.8/runpy.py", line 87, in _run_code
        exec(code, run_globals)
      File "./__main__.py", line 3, in <module>
        raise Exception('!')
    Exception: !
    error: an unhandled error occurred: Program exited with non-zero exit code: 1
 ```

### Desired behavior

Same as before, but 1 error instead of 3, drop the redundant ones.

```
Previewing update (dev)

View Live: https://app.pulumi.com/t0yv0/python-err/dev/previews/e08cb126-25ea-4c38-9640-15e381ebc32d

     Type                 Name            Plan       Info
 +   pulumi:pulumi:Stack  python-err-dev  create     *1 error*
 
Diagnostics:
  pulumi:pulumi:Stack (python-err-dev):
    error: Traceback (most recent call last):
      File "/Users/anton/.nix-profile/bin/pulumi-language-python-exec", line 107, in <module>
        loop.run_until_complete(coro)
      File "/nix/store/7q3qsarzyffcm2sjb2yhjgy0d3hzdi7y-python3-3.8.8/lib/python3.8/asyncio/base_events.py", line 616, in run_until_complete
        return future.result()
      File "/Users/anton/tmp/48h/python-err/venv/lib/python3.8/site-packages/pulumi/runtime/stack.py", line 126, in run_in_stack
        await run_pulumi_func(lambda: Stack(func))
      File "/Users/anton/tmp/48h/python-err/venv/lib/python3.8/site-packages/pulumi/runtime/stack.py", line 49, in run_pulumi_func
        func()
      File "/Users/anton/tmp/48h/python-err/venv/lib/python3.8/site-packages/pulumi/runtime/stack.py", line 126, in <lambda>
        await run_pulumi_func(lambda: Stack(func))
      File "/Users/anton/tmp/48h/python-err/venv/lib/python3.8/site-packages/pulumi/runtime/stack.py", line 149, in __init__
        func()
      File "/Users/anton/.nix-profile/bin/pulumi-language-python-exec", line 106, in <lambda>
        coro = pulumi.runtime.run_in_stack(lambda: runpy.run_path(args.PROGRAM, run_name='__main__'))
      File "/nix/store/7q3qsarzyffcm2sjb2yhjgy0d3hzdi7y-python3-3.8.8/lib/python3.8/runpy.py", line 282, in run_path
        return _run_code(code, mod_globals, init_globals,
      File "/nix/store/7q3qsarzyffcm2sjb2yhjgy0d3hzdi7y-python3-3.8.8/lib/python3.8/runpy.py", line 87, in _run_code
        exec(code, run_globals)
      File "./__main__.py", line 3, in <module>
        raise Exception('!')
    Exception: !
 ```

### Affected area/feature

Possibly Python SDK, Python language provider.

### PR DESCRIPTION
**Pull request by dixler**
# Description
This PR combines the `Program failed with an unhandled exception:` and traceback error messages and puts them into a single log message as well as omits the existing `an unhandled error occurred` message.

New State:
[asciicast](https://asciinema.org/a/GWhjkt8oeMNUvVeyb9hFtMHsF)

Current State:
[asciicast](https://asciinema.org/a/aiGJ1HomywxVkf1TAPYLK5ayq)

Input Program:
```python
"""An AWS Python Pulumi program"""

import pulumi
from pulumi_aws import s3

# Create an AWS resource (S3 Bucket)
bucket = s3.Bucket('my-bucket')
raise Exception("foobar")

# Export the name of the bucket
pulumi.export('bucket_name', bucket.id)
```

Fixes #10222 

old PR: https://github.com/pulumi/pulumi/pull/10265
## Checklist

- [ ] I have added tests that prove my fix is effective or that my feature works

- [ ] I have updated the [CHANGELOG-PENDING](https://github.com/pulumi/pulumi/blob/master/CHANGELOG_PENDING.md) file with my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Service API version

### CODE REVIEW

File: pkg/engine/lifecycletest/pulumi_test.go [hunk 1]
```diff
@@ -311,19 +311,6 @@ func TestLanguageHostDiagnostics(t *testing.T) {
 				evts []Event, res result.Result) result.Result {
 
 				assertIsErrorOrBailResult(t, res)
-				sawExitCode := false
```
Comment (t0yv0) [inline pkg/engine/lifecycletest/pulumi_test.go:314]: Why is this change needed? If I add it back the test still passes. 

AFAIK this test is testing what we call the engine (pkg/engine) and is unrelated to the Python SDK. It's trying to test that the engine would propagate the text of a returned error from the user program into a DiagEvent with:

```
sawExitCode = strings.Contains(msg, errorText) && e.Severity == diag.Error
```

It still should be the case no?

File: sdk/python/cmd/pulumi-language-python-exec [hunk 2]
```diff
@@ -23,6 +23,9 @@ It looks like the Pulumi SDK has not been installed. Have you run pip install?
 If you are running in a virtualenv, you must run pip install -r requirements.txt from inside the virtualenv.""", file=sys.stderr)
     sys.exit(1)
 
+# use exit code 32 to signal to the language host that an error message was displayed to the user
+PYTHON_PROCESS_EXITED_AFTER_SHOWING_USER_ACTIONABLE_MESSAGE_CODE = 32
```
Comment (t0yv0) [inline sdk/python/cmd/pulumi-language-python-exec:27]: 👍

File: pkg/engine/lifecycletest/pulumi_test.go (hunk 1 above)
Comment (dixler) [reply pkg/engine/lifecycletest/pulumi_test.go:314]: This is a remnant from splitting out the Golang PR. I'll add it back.

Comment (t0yv0) [reply pkg/engine/lifecycletest/pulumi_test.go:314]: Thanks! That's the only part that looks odd. The rest looks great!

**Review by t0yv0 — APPROVED** [review]
All LGTM except the lifecycletest edit.

File: CHANGELOG_PENDING.md [hunk 3]
```diff
@@ -3,6 +3,10 @@
 - [cli] Updated to the latest version of go-git.
   [#10330](https://github.com/pulumi/pulumi/pull/10330)
 
+- [sdk]
```
Comment (t0yv0) [inline CHANGELOG_PENDING.md:6]: sdk/python

Comment (t0yv0) [reply CHANGELOG_PENDING.md:6]: also provider/python

File: sdk/python/cmd/pulumi-language-python/main.go [hunk 4]
```diff
@@ -704,7 +709,15 @@ func (host *pythonLanguageHost) Run(ctx context.Context, req *pulumirpc.RunReque
 			// If the program ran, but exited with a non-zero error code.  This will happen often, since user
 			// errors will trigger this.  So, the error message should look as nice as possible.
 			if status, stok := exiterr.Sys().(syscall.WaitStatus); stok {
-				err = errors.Errorf("Program exited with non-zero exit code: %d", status.ExitStatus())
+				switch status.ExitStatus() {
+				case 0:
+					// This really shouldn't happen, but if it does, we don't want to render "non-zero exit code"
+					err = errors.Wrapf(exiterr, "Program exited unexpectedly")
```
Comment (t0yv0) [inline sdk/python/cmd/pulumi-language-python/main.go:715]: Some users will upgrade Pulumi but forget to upgrade the SDK dependency, I think that's possible. Looks like this code will be OK in this case. It is bacwards-compat.

**Pull request merged**